feat(apis_entities): copy fields when merging entities

b1rger opened the pull request:
This commit introduces a `merge_fields` method in the TempEntityClass.
It copies information for the relevant fields in TempEntityClass and can
be extended by subclasses to implement custom logic for copying custom
subclass fields.

Closes: #140

b1rger commented:
Please note that I made some decisions about how attributes are merged- i.e. the `name` field gets concatenated with a comma. There might by better approaches.

koeaw commented:
@b1rger I'm not entirely clear on the long(er)-term purpose of this seeing as we're about to drop `TempEntityClass`. Is this supposed to be moved elsewhere later on? Or can this help with _moving away_ from `TempEntityClass`?

I had a look at the linked issue but am none the wiser. It's possible I'm only missing a bit of context here.

b1rger commented:
> @b1rger I'm not entirely clear on the long(er)-term purpose of this seeing as we're about to drop `TempEntityClass`. Is this supposed to be moved elsewhere later on? Or can this help with _moving away_ from `TempEntityClass`?
> 
> I had a look at the linked issue but am none the wiser. It's possible I'm only missing a bit of context here.

At the moment when entities are merged information gets lost and there is no way for ontologies to define how the local fields of their entities should be handled when merging is done. I don't believe that `TempEntityClass` will be going away that soon, so I think it makes sense to fix this problem

sennierer commented on apis_core/apis_entities/models.py:
```diff
@@ -274,6 +274,26 @@ def get_duplicate_url(self):
+        """
+        # Inherited from RootObject
+        if self.name != other.name:
+            self.name = ", ".join([self.name, other.name])
```
I think it should be rather
```python
self.name = f"{self.name} ({other.name})"
```
for the dates `self` is overruling `other` when both objects have dates, that should be reflected in the string based fields as well

sennierer commented on apis_core/apis_entities/models.py:
```diff
@@ -274,6 +274,26 @@ def get_duplicate_url(self):
+        self.start_date_written = self.start_date_written or other.start_date_written
+        self.end_date_written = self.end_date_written or other.end_date_written
+        # we simply concatenate strings
+        self.references = "\n".join(filter(None, [self.references, other.references]))
```
see comment above. should be clear which comments stem from an object that was merged in the actual one (put in brackets or explicitly mentioned ... eg `references from merged objects:`

sennierer reviewed (changes requested):
I think the merging is currently not completely concise:
`self` is overruling `other` on some fields, others are completely equal.
I think the object that something is merged into (so `self`) should be overruling `other` and this should also be reflected in the string based fields.

Given that this should work on all classes, so include also fields that are not inheritated from `RootObject` we could refactor that to run through all fields and depending on the field type (so whether its sting, date etc. always use the same logic/function on it.

b1rger commented:
> Given that this should work on all classes, so include also fields that are not inheritated from RootObject we could refactor that to run through all fields and depending on the field type (so whether its sting, date etc. always use the same logic/function on it.

Oke, I've implemented that - the `merge_fields` method now iterates through the model fields and checks wether there is a `merge_{fieldname}` method and if so uses that. If not, it checks for a `merge_{fieldtype}` method and uses that. If no method exists and the field in `self` is empty, it copies the value of `other`.

I've implemented `merge_` methods for charfield, textfield and booleanfield. Given that `start_date_written` and `end_date_written` are parsed and the parsing would fail with the default merge method for charfields, I added extra methods for those two, demonstrating how the merge methods get chosen.

In the long run, some of the merge functionality should probably move to the RootObject or the AbstractEntity...
